core/commands/refs.go

```diff
@@ -274,43 +286,53 @@ func (rw *RefWriter) writeRefsRecursive(n ipld.Node) (int, error) {
 	return count, nil
 }
 
-func (rw *RefWriter) writeRefsSingle(n ipld.Node) (int, error) {
-	c := n.Cid()
-
-	if rw.skip(c) {
-		return 0, nil
+// visit returns two values:
+// - first indicates if we should keep traversing the DAG.
+// - second indicates if the given Cid should be printed to the user.
+//
+// visit will do branch pruning depending on rw.MaxDepth, previously visited
+// cids and whether rw.Unique is set. i.e. rw.Unique = false and
+// rw.MaxDepth = 1 disables any pruning. But setting rw.Unique to true will
+// prune already visited branches.
+func (rw *RefWriter) visit(c *cid.Cid, depth int) (bool, bool) {
+	overMaxDepth := rw.MaxDepth >= 0 && depth > rw.MaxDepth
+
+	// When not unique, only avoid exploring too deep
+	// We do not track a set of visited nodes in this case.
+	// We do not print anything too deep though.
+	if !rw.Unique {
+		return !overMaxDepth, overMaxDepth
```

Comment:
Don't we want to not explore nodes *at* max depth as well?

Reply:
Sorry, I think I misread the double negation. Right now I think it works as it should i.e.

--max-depth 1 prints just the direct children of the given CID, without fetching them (as they're links from the parent and we know we can't go deeper). Thus, items at maxdepth are never dag.Get(), but they are visited and potentially printed.